// @shopify/shopify-app-session-storage-postgresql/build/ts/postgresql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgreSQLSessionStorage = void 0;
var tslib_1 = require("tslib");
var shopify_api_1 = require("@shopify/shopify-api");
var migrations_1 = require("./migrations");
var postgres_connection_1 = require("./postgres-connection");
var postgres_migrator_1 = require("./postgres-migrator");
var defaultPostgreSQLSessionStorageOptions = {
    sessionTableName: 'shopify_sessions',
    port: 3211,
    migratorOptions: {
        migrationDBIdentifier: 'shopify_sessions_migrations',
        migrationNameColumnName: 'migration_name',
    },
};
var PostgreSQLSessionStorage = /** @class */ (function () {
    function PostgreSQLSessionStorage(dbUrl, opts) {
        if (opts === void 0) { opts = {}; }
        this.options = tslib_1.__assign(tslib_1.__assign({}, defaultPostgreSQLSessionStorageOptions), opts);
        this.internalInit = this.init(typeof dbUrl === 'string' ? dbUrl : dbUrl.toString());
        this.migrator = new postgres_migrator_1.PostgresSessionStorageMigrator(this.client, this.options.migratorOptions, migrations_1.migrationList);
        this.ready = this.migrator.applyMigrations(this.internalInit);
    }
    PostgreSQLSessionStorage.withCredentials = function (host, dbName, username, password, opts) {
        return new PostgreSQLSessionStorage(new URL("postgres://".concat(encodeURIComponent(username), ":").concat(encodeURIComponent(password), "@").concat(host, "/").concat(encodeURIComponent(dbName))), opts);
    };
    PostgreSQLSessionStorage.prototype.storeSession = function (session) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var entries, query;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        entries = session
                            .toPropertyArray()
                            .map(function (_a) {
                            var _b = tslib_1.__read(_a, 2), key = _b[0], value = _b[1];
                            return key === 'expires'
                                ? [key, Math.floor(value / 1000)]
                                : [key, value];
                        });
                        query = "\n      INSERT INTO \"".concat(this.options.sessionTableName, "\"\n      (").concat(entries.map(function (_a) {
                            var _b = tslib_1.__read(_a, 1), key = _b[0];
                            return "\"".concat(key, "\"");
                        }).join(', '), ")\n      VALUES (").concat(entries
                            .map(function (_, i) { return "".concat(_this.client.getArgumentPlaceholder(i + 1)); })
                            .join(', '), ")\n      ON CONFLICT (\"id\") DO UPDATE SET ").concat(entries
                            .map(function (_a) {
                            var _b = tslib_1.__read(_a, 1), key = _b[0];
                            return "\"".concat(key, "\" = Excluded.\"").concat(key, "\"");
                        })
                            .join(', '), ";\n    ");
                        return [4 /*yield*/, this.client.query(query, entries.map(function (_a) {
                                var _b = tslib_1.__read(_a, 2), _key = _b[0], value = _b[1];
                                return value;
                            }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.loadSession = function (id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query, rows, rawResult;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      SELECT * FROM \"".concat(this.options.sessionTableName, "\"\n      WHERE \"id\" = ").concat(this.client.getArgumentPlaceholder(1), ";\n    ");
                        return [4 /*yield*/, this.client.query(query, [id])];
                    case 2:
                        rows = _a.sent();
                        if (!Array.isArray(rows) || (rows === null || rows === void 0 ? void 0 : rows.length) !== 1)
                            return [2 /*return*/, undefined];
                        rawResult = rows[0];
                        return [2 /*return*/, this.databaseRowToSession(rawResult)];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.deleteSession = function (id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      DELETE FROM \"".concat(this.options.sessionTableName, "\"\n      WHERE \"id\" = ").concat(this.client.getArgumentPlaceholder(1), ";\n    ");
                        return [4 /*yield*/, this.client.query(query, [id])];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.deleteSessions = function (ids) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      DELETE FROM \"".concat(this.options.sessionTableName, "\"\n      WHERE \"id\" IN (").concat(ids
                            .map(function (_, i) { return "".concat(_this.client.getArgumentPlaceholder(i + 1)); })
                            .join(', '), ");\n    ");
                        return [4 /*yield*/, this.client.query(query, ids)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.findSessionsByShop = function (shop) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query, rows, results;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      SELECT * FROM \"".concat(this.options.sessionTableName, "\"\n      WHERE \"shop\" = ").concat(this.client.getArgumentPlaceholder(1), ";\n    ");
                        return [4 /*yield*/, this.client.query(query, [shop])];
                    case 2:
                        rows = _a.sent();
                        if (!Array.isArray(rows) || (rows === null || rows === void 0 ? void 0 : rows.length) === 0)
                            return [2 /*return*/, []];
                        results = rows.map(function (row) {
                            return _this.databaseRowToSession(row);
                        });
                        return [2 /*return*/, results];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.disconnect = function () {
        return this.client.disconnect();
    };
    PostgreSQLSessionStorage.prototype.init = function (dbUrl) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.client = new postgres_connection_1.PostgresConnection(dbUrl, this.options.sessionTableName);
                        return [4 /*yield*/, this.connectClient()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.createTable()];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.connectClient = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.connect()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.createTable = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        query = "\n        CREATE TABLE IF NOT EXISTS \"".concat(this.options.sessionTableName, "\" (\n          \"id\" varchar(255) NOT NULL PRIMARY KEY,\n          \"shop\" varchar(255) NOT NULL,\n          \"state\" varchar(255) NOT NULL,\n          \"isOnline\" boolean NOT NULL,\n          \"scope\" varchar(255),\n          \"expires\" integer,\n          \"onlineAccessInfo\" varchar(255),\n          \"accessToken\" varchar(255)\n        )\n      ");
                        return [4 /*yield*/, this.client.query(query)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    PostgreSQLSessionStorage.prototype.databaseRowToSession = function (row) {
        // convert seconds to milliseconds prior to creating Session object
        if (row.expires)
            row.expires *= 1000;
        return shopify_api_1.Session.fromPropertyArray(Object.entries(row));
    };
    return PostgreSQLSessionStorage;
}());
exports.PostgreSQLSessionStorage = PostgreSQLSessionStorage;
